core/crypto/abac/attributes_test.go

```diff
@@ -0,0 +1,229 @@
+
+	attribute, err := DecryptAttributeValue(attributeKey, encryptedAttribute)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if string(attribute) != expected {
+		t.Errorf("Failed decrypting attribute. Expected: %v, Actual: %v", expected, attribute)
+	}
+}
+
+func TestGetKAndValueForAttribute(t *testing.T) {
+	expected := "Software Engineer"
+
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	_, attribute, err := getKAndValueForAttribute("position", prek0, tcert)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if string(attribute) != expected {
+		t.Errorf("Failed retrieving attribute value from TCert. Expected: %v, Actual: %v", expected, string(attribute))
+	}
+}
+
+func TestGetValueForAttribute(t *testing.T) {
+	expected := "Software Engineer"
+
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	value, err := GetValueForAttribute("position", prek0, tcert)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if string(value) != expected {
+		t.Errorf("Failed retrieving attribute value from TCert. Expected: %v, Actual: %v", expected, string(value))
+	}
+}
+
+func TestParseEmptyAttributesHeader(t *testing.T) {
+	_, err := ParseAttributesHeader("")
+	if err == nil {
+		t.Error("Empty header should produce a parsing error")
+	}
+}
+
+func TestBuildAndParseAttributesHeader(t *testing.T) {
+	attributes := make(map[string]int)
+	attributes["company"] = 1
+	attributes["position"] = 2
+
+	header := string(BuildAttributesHeader(attributes)[:])
+
+	components, err := ParseAttributesHeader(header)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if len(components) != 2 {
+		t.Errorf("Error parsing header. Expecting two entries in header, found %v instead", len(components))
+	}
+
+	if components["company"] != 1 {
+		t.Errorf("Error parsing header. Expected %v with value %v, found %v instead", "company", 1, components["company"])
+	}
+
+	if components["position"] != 2 {
+		t.Errorf("Error parsing header. Expected %v with value %v, found %v instead", "position", 2, components["position"])
+	}
+}
+
+func TestReadAttributeHeader(t *testing.T) {
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	headerKey := getAttributeKey(prek0, HeaderAttributeName)
+
+	header, encrypted, err := ReadAttributeHeader(tcert, headerKey)
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !encrypted {
+		t.Errorf("Error parsing header. Expecting encrypted header.")
+	}
+
+	if len(header) != 1 {
+		t.Errorf("Error parsing header. Expecting %v entries in header, found %v instead", 1, len(header))
+	}
+
+	if header["position"] != 1 {
+		t.Errorf("Error parsing header. Expected %v with value %v, found %v instead", "position", 1, header["position"])
+	}
+}
+
+func TestReadTCertAttributeByPosition(t *testing.T) {
+	expected := "Software Engineer"
+
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	encryptedAttribute, err := ReadTCertAttributeByPosition(tcert, 1)
```

Comment:
to clarify my comment: if the function only returned the cert at position 1, this test passes and everything looks fine... but requesting the cert at position 2 would fail - and clearly that would be a bug.